ArcGISRuntimeSDKQt_CppSamples/Analysis/PerformValveIsolationTrace/PerformValveIsolationTrace.cpp

```diff
@@ -0,0 +1,273 @@
+#include "UtilityElementTraceResult.h"
+#include "UtilityNetwork.h"
+#include "UtilityNetworkDefinition.h"
+#include "UtilityNetworkSource.h"
+#include "UtilityNetworkTypes.h"
+#include "UtilityTier.h"
+#include "UtilityTraceConfiguration.h"
+#include "UtilityTraceFilter.h"
+#include "UtilityTraceParameters.h"
+
+#include <QUuid>
+
+using namespace Esri::ArcGISRuntime;
+
+namespace  {
+const QString featureServiceUrl = "https://sampleserver7.arcgisonline.com/arcgis/rest/services/UtilityNetwork/NapervilleGas/FeatureServer";
+}
+
+PerformValveIsolationTrace::PerformValveIsolationTrace(QObject* parent /* = nullptr */):
+  QObject(parent),
+  m_map(new Map(Basemap::streetsNightVector(this), this)),
+  m_startingLocationOverlay(new GraphicsOverlay(this))
+{
+  ServiceFeatureTable* distributionLineFeatureTable = new ServiceFeatureTable(featureServiceUrl + "/3", this);
+  FeatureLayer* distributionLineLayer = new FeatureLayer(distributionLineFeatureTable, this);
+
+  ServiceFeatureTable* deviceFeatureTable = new ServiceFeatureTable(featureServiceUrl + "/0", this);
+  FeatureLayer* deviceLayer = new FeatureLayer(deviceFeatureTable, this);
+
+  // add the feature layers to the map
+  m_map->operationalLayers()->append(distributionLineLayer);
+  m_map->operationalLayers()->append(deviceLayer);
+
+  m_utilityNetwork = new UtilityNetwork(featureServiceUrl, m_map, this);
+  connectSignals();
+  m_utilityNetwork->load();
+}
+
+PerformValveIsolationTrace::~PerformValveIsolationTrace() = default;
+
+void PerformValveIsolationTrace::init()
+{
+  // Register the map view for QML
+  qmlRegisterType<MapQuickView>("Esri.Samples", 1, 0, "MapView");
+  qmlRegisterType<PerformValveIsolationTrace>("Esri.Samples", 1, 0, "PerformValveIsolationTraceSample");
+}
+
+MapQuickView* PerformValveIsolationTrace::mapView() const
+{
+  return m_mapView;
+}
+
+// Set the view (created in QML)
+void PerformValveIsolationTrace::setMapView(MapQuickView* mapView)
+{
+  if (!mapView || mapView == m_mapView)
+    return;
+
+  m_mapView = mapView;
+  m_mapView->setMap(m_map);
+
+  // apply renderers
+  SimpleMarkerSymbol* startingPointSymbol = new SimpleMarkerSymbol(SimpleMarkerSymbolStyle::Cross, Qt::green, 25, this);
+  m_startingLocationOverlay->setRenderer(new SimpleRenderer(startingPointSymbol, this));
+  m_mapView->graphicsOverlays()->append(m_startingLocationOverlay);
+
+  emit mapViewChanged();
+}
+
+QStringList PerformValveIsolationTrace::categoriesList() const
+{
+  if (!m_utilityNetwork)
+    return { };
+
+  if (m_utilityNetwork->loadStatus() != LoadStatus::Loaded)
+    return { };
+
+  QList<UtilityCategory*> categories = m_utilityNetwork->definition()->categories();
+  QStringList strList;
+  for (UtilityCategory* category : categories)
+  {
+    strList << category->name();
+  }
+  return strList;
+}
+
+void PerformValveIsolationTrace::performTrace()
+{
+  if (m_selectedIndex < 0)
+    return;
+
+  // disable UI while trace is run
+  m_uiEnabled = false;
```

Comment:
maybe we can name this `runningTrace` or something? The decision about whether to disable the UI etc. is really a view cooncern.